File: gcc/testsuite/gcc.dg/loop-versioning-9.c
/* { dg-options "-O3 -fdump-tree-lversion-details" } */

/* Check that versioning can handle small groups of accesses.  */

void
f1 (int *x, int *y, int step, int n)
{
  for (int i = 0; i < n; ++i)
    x[i] = y[i * step * 2] + y[i * step * 2 + 1];
}

void
f2 (int *x, int *y, __INTPTR_TYPE__ step, int n)
{
  for (int i = 0; i < n; ++i)
    x[i] = y[i * step * 2] + y[i * step * 2 + 1];
}

void
f3 (int *x, int *y, int step, int n)
{
  for (int i = 0; i < n; ++i)
    x[i] = y[i * step * 3] + y[i * step * 3 + 2];
}

void
f4 (int *x, int *y, __INTPTR_TYPE__ step, int n)
{
  for (int i = 0; i < n; ++i)
    x[i] = y[i * step * 3] + y[i * step * 3 + 2];
}

void
f5 (int *x, int *y, int step, int n)
{
  for (int i = 0; i < n; ++i)
    x[i] = y[i * step * 4] + y[i * step * 4 + 3];
}

void
f6 (int *x, int *y, __INTPTR_TYPE__ step, int n)
{
  for (int i = 0; i < n; ++i)
    x[i] = y[i * step * 4] + y[i * step * 4 + 3];
}

/* { dg-final { scan-tree-dump-times {want to version containing loop} 6 "lversion" } } */
/* { dg-final { scan-tree-dump-times {versioned this loop} 6 "lversion" } } */

File: gcc/testsuite/gfortran.dg/loop_versioning_9.f90
! { dg-options "-O3 -fdump-tree-lversion-details" }

subroutine f1(x)
  real :: x(:, :)
  x(1, :) = 100
end subroutine f1

subroutine f2(x, i)
  real :: x(:, :)
  integer :: i
  x(i, :) = 100
end subroutine f2

subroutine f3(x)
  real :: x(:, :)
  do j = lbound(x, 2), ubound(x, 2)
     x(1, j) = 100
  end do
end subroutine f3

subroutine f4(x, i)
  real :: x(:, :)
  integer :: i
  do j = lbound(x, 2), ubound(x, 2)
     x(i, j) = 100
  end do
end subroutine f4

! { dg-final { scan-tree-dump-times {likely to be the innermost dimension} 4 "lversion" } }
! { dg-final { scan-tree-dump-not {want to version} "lversion" } }
! { dg-final { scan-tree-dump-not {versioned} "lversion" } }
